source/common/http/header_utility.cc

```diff
@@ -235,7 +235,96 @@ bool HeaderUtility::headerNameContainsUnderscore(const absl::string_view header_
   return header_name.find('_') != absl::string_view::npos;
 }
 
+namespace {
+// This function is used to validate the authority as is done in nghttp2 (based on
+// https://github.com/nghttp2/nghttp2/blob/d97bc7d8745ded136efa6e9e747f2310406893dd/lib/nghttp2_helper.c#L752).
+// Specifically, it permits the character "@" which is not permitted by oghttp2's
+// implementation. Note that this function is used for both H/1 and H/2.
```

Comment:
Can we fix oghttp2's implementation if that is incorrect?
cc @birenroy

Reply:
We had this discussion, and it is a bit nuanced.
The [authority RFC](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2) has the following structure:
```
authority   = [ userinfo "@" ] host [ ":" port ]
```

However, the [H/2 RFC](https://datatracker.ietf.org/doc/html/rfc9113#section-8.3.1) says:
```
":authority" MUST NOT include the deprecated userinfo subcomponent for "http" or "https" schemed URIs.
```
Thus, from the perspective of oghttp2, that should address H/2 headers, the validation is WAI.

Because Envoy uses the same method for both H/1 and H/2 (which is wrong and will need to be addressed), it may reject a valid H/1 authority if it uses the oghttp2 version. The desired solution is probably to separate the validation path for H/1 and H/2. Yan suggested to do this as part of the UHV work.

Reply:
So are we basically always using HTTP/1 rules for validation?  Does Balsa have equivalent checks we could use?

Reply:
I couldn't find something relevant at: https://github.com/google/quiche/tree/main/quiche/balsa.
cc @bnc-google who may know if Balsa has H/1 authority validator.

Reply:
No, there is no authority validator in balsa_parser.cc. It validates URL and method and HTTP version, but not authority.

Reply:
Alas.  Ok done with my drive by, thanks for your patience @adisuissa ,  I'll leave the rest to Yan. :-)